File: ios/ContractRunner/Sources/ContractRunner/CatalogUrlBuilder.swift
import Foundation

public struct CatalogFilter: Equatable {
    public let key: String
    public let value: String

    public init(key: String, value: String) {
        self.key = key
        self.value = value
    }
}

public func buildCatalogUrls(
    baseUrl: String,
    encodedQuery: String? = nil,
    mediaType: String,
    catalogId: String,
    skip: Int = 0,
    limit: Int = 20,
    filters: [CatalogFilter] = []
) -> [String] {
    let normalizedBaseUrl = baseUrl.trimmingCharacters(in: .whitespacesAndNewlines).trimmingTrailingSlashes()
    let normalizedMediaType = mediaType.trimmingCharacters(in: .whitespacesAndNewlines).lowercased()
    let normalizedCatalogId = catalogId.trimmingCharacters(in: .whitespacesAndNewlines)
    let normalizedSkip = max(skip, 0)
    let normalizedLimit = max(limit, 1)
    let normalizedQuery = normalizeEncodedQuery(encodedQuery)
    let normalizedFilters = normalizeCatalogFilters(filters)

    let pathRoot = normalizedBaseUrl + "/catalog/" + encodePathSegment(normalizedMediaType) + "/" + encodePathSegment(normalizedCatalogId)
    var urls: [String] = []

    if normalizedSkip == 0 && normalizedFilters.isEmpty {
        urls.append(pathRoot + ".json" + querySuffix(normalizedQuery))
    }

    var pathExtras: [String] = [
        "skip=\(encodePathSegment(String(normalizedSkip)))",
        "limit=\(encodePathSegment(String(normalizedLimit)))"
    ]
    pathExtras.append(contentsOf: normalizedFilters.map { "\(encodePathSegment($0.key))=\(encodePathSegment($0.value))" })
    urls.append(pathRoot + "/" + pathExtras.joined(separator: "/") + ".json" + querySuffix(normalizedQuery))

    var queryParts: [String] = []
    if let normalizedQuery {
        queryParts.append(normalizedQuery)
    }
    queryParts.append("skip=\(encodeQueryComponent(String(normalizedSkip)))")
    queryParts.append("limit=\(encodeQueryComponent(String(normalizedLimit)))")
    queryParts.append(contentsOf: normalizedFilters.map { "\(encodeQueryComponent($0.key))=\(encodeQueryComponent($0.value))" })
    urls.append(pathRoot + ".json?" + queryParts.joined(separator: "&"))

    return urls
}

private func normalizeCatalogFilters(_ filters: [CatalogFilter]) -> [CatalogFilter] {
    return filters
        .compactMap { filter in
            let key = filter.key.trimmingCharacters(in: .whitespacesAndNewlines)
            let value = filter.value.trimmingCharacters(in: .whitespacesAndNewlines)
            guard !key.isEmpty, !value.isEmpty else {
                return nil
            }
            return CatalogFilter(key: key, value: value)
        }
        .sorted {
            let leftKey = $0.key.lowercased()
            let rightKey = $1.key.lowercased()
            if leftKey != rightKey {
                return leftKey < rightKey
            }
            return $0.value.lowercased() < $1.value.lowercased()
        }
}

private func normalizeEncodedQuery(_ value: String?) -> String? {
    guard let trimmed = value?.trimmingCharacters(in: .whitespacesAndNewlines), !trimmed.isEmpty else {
        return nil
    }
    let normalized = trimmed.hasPrefix("?") ? String(trimmed.dropFirst()) : trimmed
    return normalized.isEmpty ? nil : normalized
}

private func querySuffix(_ encodedQuery: String?) -> String {
    guard let encodedQuery else {
        return ""
    }
    return "?\(encodedQuery)"
}

private func encodePathSegment(_ value: String) -> String {
    return encodeQueryComponent(value)
}

private func encodeQueryComponent(_ value: String) -> String {
    let allowed = CharacterSet.alphanumerics.union(CharacterSet(charactersIn: "-._~"))
    return value.addingPercentEncoding(withAllowedCharacters: allowed) ?? value
}

private extension String {
    func trimmingTrailingSlashes() -> String {
        var result = self
        while result.hasSuffix("/") {
            result.removeLast()
        }
        return result
    }
}

File: android/core-domain/src/main/kotlin/com/crispy/tv/domain/catalog/CatalogUrlBuilder.kt
package com.crispy.tv.domain.catalog

import java.net.URLEncoder
import java.nio.charset.StandardCharsets
import java.util.Locale

fun buildCatalogUrls(
    baseUrl: String,
    encodedQuery: String? = null,
    mediaType: String,
    catalogId: String,
    skip: Int = 0,
    limit: Int = 20,
    filters: List<CatalogFilter> = emptyList(),
): List<String> {
    val normalizedBaseUrl = baseUrl.trim().trimEnd('/')
    val normalizedMediaType = mediaType.trim().lowercase(Locale.US)
    val normalizedCatalogId = catalogId.trim()
    val normalizedSkip = skip.coerceAtLeast(0)
    val normalizedLimit = limit.coerceAtLeast(1)
    val normalizedQuery = encodedQuery.normalizedEncodedQuery()
    val normalizedFilters = normalizeCatalogFilters(filters)

    val pathRoot = buildString {
        append(normalizedBaseUrl)
        append("/catalog/")
        append(encodePathSegment(normalizedMediaType))
        append('/')
        append(encodePathSegment(normalizedCatalogId))
    }

    val urls = mutableListOf<String>()
    if (normalizedSkip == 0 && normalizedFilters.isEmpty()) {
        urls += "$pathRoot.json${querySuffix(normalizedQuery)}"
    }

    val pathExtras = buildList {
        add("skip=${encodePathSegment(normalizedSkip.toString())}")
        add("limit=${encodePathSegment(normalizedLimit.toString())}")
        normalizedFilters.forEach { filter ->
            add("${encodePathSegment(filter.key)}=${encodePathSegment(filter.value)}")
        }
    }
    urls += "$pathRoot/${pathExtras.joinToString("/")}.json${querySuffix(normalizedQuery)}"

    val queryParts = mutableListOf<String>()
    normalizedQuery?.let(queryParts::add)
    queryParts += "skip=${encodeQueryComponent(normalizedSkip.toString())}"
    queryParts += "limit=${encodeQueryComponent(normalizedLimit.toString())}"
    normalizedFilters.forEach { filter ->
        queryParts += "${encodeQueryComponent(filter.key)}=${encodeQueryComponent(filter.value)}"
    }
    urls += "$pathRoot.json?${queryParts.joinToString("&")}"

    return urls
}

private fun normalizeCatalogFilters(filters: List<CatalogFilter>): List<CatalogFilter> {
    return filters
        .mapNotNull { filter ->
            val key = filter.key.trim()
            val value = filter.value.trim()
            if (key.isEmpty() || value.isEmpty()) {
                null
            } else {
                CatalogFilter(key = key, value = value)
            }
        }
        .sortedWith(compareBy<CatalogFilter>({ it.key.lowercase(Locale.US) }, { it.value.lowercase(Locale.US) }))
}

private fun String?.normalizedEncodedQuery(): String? {
    val normalized = this?.trim().orEmpty().removePrefix("?")
    return normalized.ifEmpty { null }
}

private fun querySuffix(encodedQuery: String?): String {
    return encodedQuery?.let { "?$it" } ?: ""
}

private fun encodePathSegment(value: String): String {
    return encodeQueryComponent(value)
}

private fun encodeQueryComponent(value: String): String {
    return URLEncoder.encode(value, StandardCharsets.UTF_8).replace("+", "%20")
}
